src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,684 @@
+  }
+  if (finish_op_ctx == nullptr) {
+    lderr(m_cct) << "lost exclusive lock" << dendl;
+    finish(-EROFS);
+    return;
+  }
+
+  auto ctx = new FunctionContext([this, finish_op_ctx](int r) {
+      handle_write_object(r);
+      finish_op_ctx->complete(0);
+    });
+  librados::AioCompletion *comp = create_rados_callback(ctx);
+  int r = m_dst_io_ctx.aio_operate(m_dst_oid, comp, &op, dst_snap_seq,
+                                   dst_snap_ids);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::handle_write_object(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  if (r == -ENOENT) {
+    r = 0;
+  }
+  if (r < 0) {
+    lderr(m_cct) << "failed to write to destination object: " << cpp_strerror(r)
+                 << dendl;
+    finish(r);
+    return;
+  }
+
+  m_write_ops.erase(m_write_ops.begin());
+  if (!m_write_ops.empty()) {
+    send_write_object();
+    return;
+  }
+
+  send_update_object_map();
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::send_update_object_map() {
+  m_dst_image_ctx->owner_lock.get_read();
+  m_dst_image_ctx->snap_lock.get_read();
+  if (!m_dst_image_ctx->test_features(RBD_FEATURE_OBJECT_MAP,
+                                      m_dst_image_ctx->snap_lock) ||
+      m_dst_object_state.empty()) {
+    m_dst_image_ctx->snap_lock.put_read();
+    m_dst_image_ctx->owner_lock.put_read();
+    finish(0);
+    return;
+  } else if (m_dst_image_ctx->object_map == nullptr) {
+    // possible that exclusive lock was lost in background
+    lderr(m_cct) << "object map is not initialized" << dendl;
+
+    m_dst_image_ctx->snap_lock.put_read();
+    m_dst_image_ctx->owner_lock.put_read();
+    finish(-EINVAL);
+    return;
+  }
+
+  assert(m_dst_image_ctx->object_map != nullptr);
+
+  auto &dst_object_state = *m_dst_object_state.begin();
+  auto it = m_snap_map.find(dst_object_state.first);
+  assert(it != m_snap_map.end());
+  auto dst_snap_id = it->second.front();
+  auto object_state = dst_object_state.second;
+  m_dst_object_state.erase(m_dst_object_state.begin());
+
+  ldout(m_cct, 20) << "dst_snap_id=" << dst_snap_id << ", object_state="
+                   << static_cast<uint32_t>(object_state) << dendl;
+
+  auto finish_op_ctx = start_lock_op(m_dst_image_ctx->owner_lock);
+  if (finish_op_ctx == nullptr) {
+    lderr(m_cct) << "lost exclusive lock" << dendl;
+    m_dst_image_ctx->snap_lock.put_read();
+    m_dst_image_ctx->owner_lock.put_read();
+    finish(-EROFS);
+    return;
+  }
+
+  auto ctx = new FunctionContext([this, finish_op_ctx](int r) {
+      handle_update_object_map(r);
+      finish_op_ctx->complete(0);
+    });
+
+  RWLock::WLocker object_map_locker(m_dst_image_ctx->object_map_lock);
+  bool sent = m_dst_image_ctx->object_map->template aio_update<
+    Context, &Context::complete>(dst_snap_id, m_dst_object_number, object_state,
+                                 {}, {}, ctx);
+  assert(sent);
```

Comment:
Nit: if ```dst_snap_id == CEPH_NOSNAP``` and the HEAD object doesn't exist, it may be possible that the in-memory object map is already at NONEXISTENT and thus ```sent``` will be false